common/rate-limiter/src/async_lib.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::rate_limit::{Bucket, SharedBucket};
+use diem_infallible::Mutex;
+use futures::{
+    io::{AsyncRead, Error},
+    ready,
+    task::{Context, Poll},
+    AsyncWrite, Future,
+};
+use pin_project::pin_project;
+use std::{io, pin::Pin, sync::Arc};
+use tokio::time::{delay_until, Delay};
+
+/// An inner struct for keeping track of the delay, and bucket of rate limiting
+struct PollRateLimiter {
+    bucket: SharedBucket,
+    delay: Option<Delay>,
+}
+
+impl PollRateLimiter {
+    fn new(bucket: Option<SharedBucket>) -> Self {
+        let bucket =
+            bucket.unwrap_or_else(|| Arc::new(Mutex::new(Bucket::open(String::from("None")))));
+        PollRateLimiter {
+            bucket,
+            delay: None,
+        }
+    }
+
+    /// Poll and attempt to acquire the `requested` amount of tokens.
+    /// Keep trying until some amount of tokens are acquired.  Note: This doesn't provide
+    /// fairness so if two pollers hold the same bucket, one could continually lose.
+    fn poll_acquire(&mut self, cx: &mut Context<'_>, requested: usize) -> Poll<usize> {
```

Comment:
I know why this bothered me now... there's no logging and it is an infinite loop that makes a deep assumption about the bucket implementation -- that is that the bucket is nonzero

can we perhaps make a max iterations in the loop? (in another PR)

Reply:
We could make max iterations on the loop in another PR.

Reply:
is the loop really an issue? it's like polling a lock with a delay if it's already held

```rust
loop {
    if let Some(permit) = lock.try_acquire() {
        return permit;
    } else {
        sleep(delay);
    }
}
```

if we're concerned with starvation, then i'd agree; we don't currently guarantee any fairness across waiters. of course, for the current network use-case where waiters are all from the same IP / client, I don't think fairness is that important. we should definitely document the lack of fairness somewhere clearly though : )